### PR TITLE
Structure-structure coupling

### PR DESCRIPTION
**Pull request by Alex-tru**
This extension of the adapter allows to read displacements and write forces to calculix participants. Should maintain unaltered FSI functionality, though it has not been tested yet.

### CODE REVIEW

**Comment by arusch** [comment]
Hi Àlex,

thank you for this pull request. I will review the changes as quickly as possible but probably not before the end of this week.

Cheers
Alexander

**Comment by Alex-tru** [comment]
Ok, thanks in advance!

**Comment by Alex-tru** [comment]
Hi Alexander,
thanks a lot for the tips and recommendations. Let me comment on the points you mentioned:
1. Restored.
2, 3, 4. I think that most for loops with different style were the ones for debug-output, which I've removed now. For the rest, style should now be fitting in.
5. No, there's no reason. A comment-typo.
6. This has a reason. In `CCXHelpers.c`, in the function `getXbounIndices` I introduced a switch statement using the `CouplingDataType` as case variable. It is not strictly necessary, but otherwise an additional flag has to be used.

File: Makefile [hunk 1]
```diff
@@ -106,4 +106,4 @@ $(OBJDIR):
 	mkdir -p $(OBJDIR)
 
 clean:
-	rm -f $(OBJDIR)/*.o $(OBJDIR)/ccx_2.13.a $(OBJDIR)/ccx_preCICE
+rm -f $(OBJDIR)/*.o $(OBJDIR)/ccx_2.13.a $(OBJDIR)/ccx_preCICE
```
Comment (arusch) [inline Makefile]: Watch out: Indentation in `Makefiles` is crucial. Deleting the tab here creates a `Makefile` that is not functioning. Please redo this such that there are no changes at the `Makefile` at all.

File: adapter/PreciceInterface.c [hunk 2]
```diff
@@ -188,18 +187,18 @@ void Precice_ReadCouplingData( SimulationData * sim )
...
-                break;
-	    case DISPLACEMENTDELTAS:
+				// Read and set displacements as single point constraints (Dirichlet BC)
+				precicec_readBlockVectorData( interfaces[i]->displacementsDataID, interfaces[i]->numNodes, interfaces[i]->preciceNodeIDs, interfaces[i]->nodeVectorData );
+				setNodeDisplacements( interfaces[i]->nodeVectorData, interfaces[i]->numNodes, interfaces[i]->xbounIndices, sim->xboun );
+				break;
+	    	case DISPLACEMENTDELTAS:
                 printf( "DisplacementDeltas cannot be used as read data\n" );
-                fflush( stdout );
+				fflush( stdout );
```
Comment (arusch) [inline adapter/PreciceInterface.c]: White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

File: adapter/PreciceInterface.c [hunk 3]
```diff
@@ -188,18 +187,18 @@ void Precice_ReadCouplingData( SimulationData * sim )
...
-                setNodeForces( interfaces[i]->nodeVectorData, interfaces[i]->numNodes, interfaces[i]->xforcIndices, sim->xforc );
+                setNodeForces( interfaces[i]->preciceNodeIDs, interfaces[i]->nodeVectorData, interfaces[i]->numNodes, interfaces[i]->xforcIndices, sim->xforc);
                 break;
             case DISPLACEMENTS:
-                printf( "Displacements cannot be used as read data\n" );
-                fflush( stdout );
-                exit( EXIT_FAILURE );
-                break;
-	    case DISPLACEMENTDELTAS:
+				// Read and set displacements as single point constraints (Dirichlet BC)
```
Comment (arusch) [inline adapter/PreciceInterface.c]: White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

File: adapter/PreciceInterface.c [hunk 4]
```diff
@@ -280,14 +279,13 @@ void Precice_WriteCouplingData( SimulationData * sim )
                 getNodeDisplacements( interfaces[i]->nodeIDs, interfaces[i]->numNodes, sim->vold, sim->mt, interfaces[i]->nodeVectorData );
                 precicec_writeBlockVectorData( interfaces[i]->displacementsDataID, interfaces[i]->numNodes, interfaces[i]->preciceNodeIDs, interfaces[i]->nodeVectorData );
                 break;
-	    case DISPLACEMENTDELTAS:
+	    	case DISPLACEMENTDELTAS:
```
Comment (arusch) [inline adapter/PreciceInterface.c]: White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

File: adapter/PreciceInterface.c [hunk 5]
```diff
@@ -330,13 +328,13 @@ void PreciceInterface_Create( PreciceInterface * interface, SimulationData * sim
 	interface->preciceNodeIDs = NULL;
 	interface->triangles = NULL;
 	interface->nodeScalarData = NULL;
-        interface->nodeVectorData = NULL;
+    interface->nodeVectorData = NULL;
```
Comment (arusch) [inline adapter/PreciceInterface.c]: White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

File: adapter/PreciceInterface.c [hunk 6]
```diff
@@ -469,17 +467,26 @@ void PreciceInterface_ConfigureCouplingData( PreciceInterface * interface, Simul
...
+            printf( "Read data '%s' found.\n", config->readDataNames[i] );
+            break;
+        }
+        else if ( strcmp1( config->readDataNames[i], "Displacements" + i ) == 0 )
+        {
+        	PreciceInterface_EnsureValidNodesMeshID( interface );
+        	interface->readData = DISPLACEMENTS;
+			interface->xbounIndices = malloc( interface->numNodes * 3 * sizeof( int ) );
+            interface->displacementsDataID = precicec_getDataID( config->readDataNames[i], interface->nodesMeshID );
+			getXbounIndices( interface->nodeIDs, interface->numNodes, sim->nboun, sim->ikboun, sim->ilboun, interface->xbounIndices, DISPLACEMENTS );
```
Comment (arusch) [inline adapter/PreciceInterface.c]: White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

File: adapter/PreciceInterface.c [hunk 7]
```diff
@@ -515,20 +522,27 @@ void PreciceInterface_ConfigureCouplingData( PreciceInterface * interface, Simul
...
 			printf( "Write data '%s' found.\n", config->writeDataNames[i] );
 		}
-                else if ( strcmp1( config->writeDataNames[i], "Displacements" + i ) == 0 )
-                {
-                        PreciceInterface_EnsureValidNodesMeshID( interface );
-            		interface->writeData = DISPLACEMENTS;
-                        interface->displacementsDataID = precicec_getDataID( config->writeDataNames[i], interface->nodesMeshID );
-                        printf( "Write data '%s' found.\n", config->writeDataNames[i] );
-                }
+        else if ( strcmp1( config->writeDataNames[i], "Displacements" + i ) == 0 )
```
Comment (arusch) [inline adapter/PreciceInterface.c]: White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

File: adapter/PreciceInterface.c [hunk 8]
```diff
@@ -587,9 +601,8 @@ void PreciceInterface_FreeData( PreciceInterface * preciceInterface )
 		free( preciceInterface->xloadIndices );
 	}
     
-    	if ( preciceInterface->xforcIndices != NULL ) {
-        	free( preciceInterface->xforcIndices );
-    	}
+    if ( preciceInterface->xforcIndices != NULL )
```
Comment (arusch) [inline adapter/PreciceInterface.c]: White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

File: adapter/PreciceInterface.h [hunk 9]
```diff
@@ -23,7 +23,7 @@
  *  Displacements - FSI data to be written (by the Calculix adapter)
  *  DisplacementDeltas - FSI data to be written (by the Calculix adapter)
  */
-enum CouplingDataType {TEMPERATURE, HEAT_FLUX, CONVECTION, FORCES, DISPLACEMENTS, DISPLACEMENTDELTAS};
+//enum CouplingDataType {TEMPERATURE, HEAT_FLUX, CONVECTION, FORCES, DISPLACEMENTS, DISPLACEMENTDELTAS};
```
Comment (arusch) [inline adapter/PreciceInterface.h]: Your change concerning the position of the `CouplingDataType` enumeration is confirmed, but please also delete the deprecated code here and don't leave it commented out.

File: adapter/PreciceInterface.h [hunk 10]
```diff
@@ -55,7 +55,7 @@ typedef struct PreciceInterface {
 
 	// Arrays to store the coupling data
 	double * nodeScalarData;
-        double * nodeVectorData; //Forces, displacements and displacementDeltas are vector quantities
+    double * nodeVectorData; //Forces, displacements and displacementDeltas are vector quantities
```
Comment (arusch) [inline adapter/PreciceInterface.h]: White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

File: adapter/PreciceInterface.h [hunk 11]
```diff
@@ -101,11 +101,11 @@ typedef struct SimulationData {
 	char * set;
 	double * co;
 	ITG nboun;
-        ITG nforc; //total number of forces
+    ITG nforc; //total number of forces
```
Comment (arusch) [inline adapter/PreciceInterface.h]: White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

File: nonlingeo_precice.c [hunk 12]
```diff
@@ -221,6 +221,7 @@ void nonlingeo_precice(double **cop, ITG *nk, ITG **konp, ITG **ipkonp, char **l
       .xboun = xboun,
       .ntmat_ = ntmat_,
       .vold = vold,
+	  .fn = fn,
```
Comment (arusch) [inline nonlingeo_precice.c]: White spaces seem to have been mixed up a little. Please correct this (at several positions in the code).

**Pull request merged**